### PR TITLE
Implement receive of messages, deleted history messages, configurable server, typing status

### PR DESCRIPTION
**Pull request by a-andreyev**
Contributes to #13, #5, #8

### CODE REVIEW

File: src/messageschannel.cpp [hunk 1]
```diff
@@ -152,26 +152,33 @@ void MatrixMessagesChannel::processMessageEvent(const QMatrixClient::RoomMessage
...
         header[QStringLiteral("message-sender-id")] = QDBusVariant(m_connection->selfID());
     } else {
-        header[QStringLiteral("message-sender")]   = QDBusVariant(m_connection->ensureContactHandle(event->senderId()));
+        header[QStringLiteral("message-sender")]    = QDBusVariant(m_connection->ensureContactHandle(event->senderId()));
         header[QStringLiteral("message-sender-id")] = QDBusVariant(event->senderId());
     }
+    header[QStringLiteral("message-sent")]  = QDBusVariant(event->timestamp().toMSecsSinceEpoch() / 1000);
+    header[QStringLiteral("message-received")] = QDBusVariant(event->timestamp().toMSecsSinceEpoch() / 1000);
+    if (event->isRedacted()) {
+        header[QStringLiteral("message-type")]    = QDBusVariant(Tp::ChannelTextMessageTypeDeliveryReport);
```
Comment (Kaffeine) [inline src/messageschannel.cpp]: A delivery report is just a 'report' and does not present a message. Probably we want the type to be 'normal' to make the client add the message as an incoming or outgoing message though it is already known to be deleted.

The next part would be to implement deleted messages on the client (UI) side.

File: src/messageschannel.cpp [hunk 2]
```diff
@@ -152,26 +152,33 @@ void MatrixMessagesChannel::processMessageEvent(const QMatrixClient::RoomMessage
...
+        header[QStringLiteral("delivery-status")] = QDBusVariant(Tp::DeliveryStatusDeleted);
+        header[QStringLiteral("delivery-token")]  = QDBusVariant(event->transactionId());
```
Comment (Kaffeine) [inline src/messageschannel.cpp]: Delivery token is [an identifier for the message to which this delivery report refers](https://telepathy.freedesktop.org/spec/Channel_Interface_Messages.html#Simple-Type:Delivery_Report_Header_Key). It is valid only for a delivery report. In this case the message (should be) a normal message with known delivery status. It does not refer to anything. And the last issue is that transactionId() returns an empty string, but the delivery-token must not be empty.

File: src/messageschannel.cpp [hunk 3]
```diff
@@ -152,26 +152,33 @@ void MatrixMessagesChannel::processMessageEvent(const QMatrixClient::RoomMessage
...
+    }
+    else {
```
Comment (Kaffeine) [inline src/messageschannel.cpp]: I would suggest reformating the code to `} else {`, but in this case we can just drop this part of the patch.

File: src/connection.cpp [hunk 4]
```diff
@@ -491,6 +490,16 @@ void MatrixConnection::onSyncDone()
...
 
     m_contactListIface->setContactListState(Tp::ContactListStateSuccess);
+
+    // Found at KitsuneRal's code in Quaternion:
+    // Borrowed the logic from Quiark's code in Tensor to cache not too
+    // aggressively and not on the first sync. The static variable instance
+    // is created per-closure, meaning per-connection (which is why this
+    // code is not in MessagesChannel).
+    m_connection->sync(30*1000); // TODO: get rid of magic number
+    static int counter = 0;
```
[the remark this answers is not in the document]
Comment (a-andreyev) [reply src/connection.cpp]: Got it! Agree

File: src/messageschannel.cpp (hunk 1 above)
Comment (a-andreyev) [reply src/messageschannel.cpp]: thank you! propably I have same issue with `MatrixMessagesChannel::onPendingEventChanged`

File: src/messageschannel.cpp (hunk 2 above)
Comment (a-andreyev) [reply src/messageschannel.cpp]: thank you! propably I have same issue with `MatrixMessagesChannel::onPendingEventChanged`

File: src/connection.cpp [hunk 5]
```diff
@@ -491,6 +510,7 @@ void MatrixConnection::onSyncDone()
     }
 
     m_contactListIface->setContactListState(Tp::ContactListStateSuccess);
+    m_connection->sync();
```
Comment (Kaffeine) [inline src/connection.cpp]: Why this is needed? It looks wrong (and maybe traffic-expensive) to sync again on sync done.
Do we need to sync if we're online? Maybe the thing is that it is in the Matrix (or QMatrixClient) nature to lose the connection during a session and connect again and again. I hope that in this case, we can request sync on the right event — "on connection established" or something like that.

Comment (a-andreyev) [reply src/connection.cpp]: contributed to libqmatrixclient https://github.com/QMatrixClient/libqmatrixclient/pull/274
version 0.5 of the library have `syncLoop` method

File: src/messageschannel.cpp (hunk 2 above)
Comment (a-andreyev) [reply src/messageschannel.cpp]: changed from `transactionId` to `id` at `MatrixMessagesChannel::onPendingEventChanged`

**Comment by a-andreyev** [comment]
thank you for the help, probably ready for a new review

**Comment by a-andreyev** [comment]
added #17

**Comment by a-andreyev** [comment]
added #15 and #10

File: src/messageschannel.hpp [hunk 6]
```diff
@@ -48,6 +49,7 @@ class JoinRoomJob;
 class UploadContentJob;
 class GetContentJob;
 class DownloadFileJob;
+class SetTypingJob;
```
Comment (Kaffeine) [inline src/messageschannel.hpp]: Not used?

File: src/messageschannel.cpp [hunk 7]
```diff
@@ -185,3 +202,47 @@ void MatrixMessagesChannel::fetchHistory()
...
+        m_chatStateIface->chatStateChanged(handle, Tp::ChannelChatStateComposing);
+    }
+}
+
+void MatrixMessagesChannel::reactivateLocalTyping()
+{
+    sendTypingNotification(true);
+}
+
+void MatrixMessagesChannel::setChatStateCallback(uint state, Tp::DBusError *error)
```
Comment (Kaffeine) [inline src/messageschannel.cpp]: I think this `Callback` word actually confuses a reader.
What I would expect from a "set chat state callback" method? I would expect a setter for some 'chat state' callback.
On the other hand this method actually set chat state (in Matrix backend). IMO `setChatState` sounds good as it is and it actually do what the name means. It is a coincide that the method can be used as a callback for the interface.

File: src/messageschannel.hpp (hunk 6 above)
Comment (a-andreyev) [reply src/messageschannel.hpp]: If I'm not mistaken, I'm using it at `MatrixMessagesChannel::sendTypingNotification`: `callApi<QMatrixClient::SetTypingJob>`

File: src/messageschannel.cpp [hunk 8]
```diff
@@ -119,12 +123,19 @@ void MatrixMessagesChannel::onPendingEventChanged(int pendingEventIndex)
...
     header[QStringLiteral("message-type")]      = QDBusVariant(Tp::ChannelTextMessageTypeDeliveryReport);
     header[QStringLiteral("delivery-status")]   = QDBusVariant(tpDeliveryStatus);
-    header[QStringLiteral("delivery-token")]    = QDBusVariant(pendingEvent.event()->transactionId());
+    header[QStringLiteral("delivery-token")]    = QDBusVariant(pendingEvent.event()->id());
     partList << header;
 
     addReceivedMessage(partList);
 }
 
+void MatrixMessagesChannel::sendTypingNotification(bool isTyping)
```
Comment (Kaffeine) [inline src/messageschannel.cpp]: I would suggest to wire Tp::ChannelChatState down to this interface so we'll have less to refactor once Matrix will gain something more than trivial `bool` state support.
Name `sendChatStateNotification()` would fit the new signature. :)

File: src/messageschannel.hpp [hunk 9]
```diff
@@ -87,6 +99,8 @@ class MatrixMessagesChannel : public Tp::BaseChannelTextType
     Tp::BaseChannelGroupInterfacePtr m_groupIface;
     Tp::BaseChannelRoomInterfacePtr m_roomIface;
     Tp::BaseChannelRoomConfigInterfacePtr m_roomConfigIface;
+
+    QTimer *m_localTypingTimer;
```
Comment (Kaffeine) [inline src/messageschannel.hpp]:
```cpp
QTimer *m_localTypingTimer = nullptr;
```

File: src/messageschannel.cpp [hunk 10]
```diff
@@ -30,14 +30,17 @@
...
 
 MatrixMessagesChannel::MatrixMessagesChannel(MatrixConnection *connection, QMatrixClient::Room *room, Tp::BaseChannel *baseChannel)
     : Tp::BaseChannelTextType(baseChannel),
       m_connection(connection),
       m_room(room),
       m_targetHandle(baseChannel->targetHandle()),
       m_targetHandleType(baseChannel->targetHandleType()),
-      m_targetId(baseChannel->targetID())
+      m_targetId(baseChannel->targetID()),
+      m_localTypingTimer(0)
```
Comment (Kaffeine) [inline src/messageschannel.cpp]: Please use 'in class definition' initialization instead. (I know I have (or had?) such one in Morse but let's avoid it in new code.)

File: src/messageschannel.hpp (hunk 6 above)
Comment (a-andreyev) [reply src/messageschannel.hpp]: oh, my bad, you're right

File: src/messageschannel.cpp [hunk 11]
```diff
@@ -152,26 +163,32 @@ void MatrixMessagesChannel::processMessageEvent(const QMatrixClient::RoomMessage
     qDebug().noquote() << Q_FUNC_INFO << "Process message" << doc.toJson(QJsonDocument::Indented);
     Tp::MessagePart header;
     header[QStringLiteral("message-token")] = QDBusVariant(event->id());
-    header[QStringLiteral("message-sent")]  = QDBusVariant(event->timestamp().toMSecsSinceEpoch() / 1000);
```
Comment (Kaffeine) [inline src/messageschannel.cpp]: This looks like move for the movement itself. The only justified move is 'message-type' IMO.
The reformat violates Qt coding style but meh, let's keep it if you wish :).

File: src/messageschannel.cpp [hunk 12]
```diff
@@ -185,3 +202,47 @@ void MatrixMessagesChannel::fetchHistory()
...
+
+    if (!m_localTypingTimer) {
+        m_localTypingTimer = new QTimer(this);
+        m_localTypingTimer->setInterval(5000); // FIXME: magic number
+        connect(m_localTypingTimer, &QTimer::timeout, this, &MatrixMessagesChannel::reactivateLocalTyping);
+    }
+
+    bool isTyping = false;
+    if (state == Tp::ChannelChatStateComposing) {
+        // m_api->setMessageAction(m_targetPeer, TelegramNamespace::MessageActionTyping);
```
Comment (Kaffeine) [inline src/messageschannel.cpp]: Please remove the comment

File: src/messageschannel.cpp (hunk 11 above)
Comment (a-andreyev) [reply src/messageschannel.cpp]: I'm actually really into follow your advice and follow Qt coding style :+1: Sorry for current state :confused: 
I'm ready to mark it as resolved currently but add to some todo and try [uncrustify](https://gist.github.com/a-andreyev/87033eb72696afedbee6bcbf83fb2def)

Comment (a-andreyev) [reply src/messageschannel.cpp]: removed fancy spaces at `sendMessage` part

Comment (Kaffeine) [reply src/messageschannel.cpp]: You still moved 'sent' and 'received' keys. I like the decision to move the 'type' key next to received though.

File: src/messageschannel.cpp [hunk 13]
```diff
@@ -119,18 +122,26 @@ void MatrixMessagesChannel::onPendingEventChanged(int pendingEventIndex)
...
 
     addReceivedMessage(partList);
 }
 
+void MatrixMessagesChannel::sendChatStateNotification(uint state)
+{
+    m_room->connection()->
+            callApi<QMatrixClient::SetTypingJob>
+            (QMatrixClient::BackgroundRequest,
+             m_connection->matrix()->user()->id(), m_room->id(), (state == Tp::ChannelChatStateComposing), 10000); // FIXME: magic number
```
Comment (Kaffeine) [inline src/messageschannel.cpp]: The last issue to solve are the magic numbers and then we can accept this :+1: as a big step forward :).

I would suggest to add
```
constexpr int c_repeatInterval = 5000;
```
and use it for the timer.

As for about the SetTypingJob, the timeout is optional so maybe we can drop the argument for now?

Comment (a-andreyev) [reply src/messageschannel.cpp]: Thank you!
I'm not sure about `c_chatStateResendInterval` apropriate place and leaved it right before `setInterval` method

File: src/messageschannel.hpp [hunk 14]
```diff
@@ -61,17 +62,27 @@ class MatrixMessagesChannel : public Tp::BaseChannelTextType
...
-    QString sendMessageCallback(const Tp::MessagePartList &messageParts, uint flags, Tp::DBusError *error);
+    QString sendMessage(const Tp::MessagePartList &messageParts, uint flags, Tp::DBusError *error);
     void processMessageEvent(const QMatrixClient::RoomMessageEvent *event);
 
     void messageAcknowledgedCallback(const QString &messageId);
 
     void fetchHistory();
 
+public slots:
+    void onTypingChanged();
```
Comment (Kaffeine) [inline src/messageschannel.hpp]: Please group all handler together, namely `onTypingChanged` should go after `onPendingEventChanged`.
It is hard to decide where to put the methods but a 'section per method' is definitely overkilling :smile:.

Let's keep in 'public' methods that can be called from the outside without Matrix or Telepathy side confusing.
E.g. sendMessage() that accepts Telepathy MessagePartList can be called from anywhere and will do what it is means to do.

I also read the spec and found that setChatState (here and in Morse) should also set own chat state in the interface.
The interface should be changed in TelepathyQt first but we can make setChatState public with such (future) public correctness in mind.

Comment (a-andreyev) [reply src/messageschannel.hpp]: Tried to follow proposed changes: hidden not only non needed `public slots` section but also `protected slots`.
And also commented out not implemented `messageAcknowledged` (that should be changed  to `messageAcknowledged`)

**Pull request merged**